wgsl/index.bs

```diff
@@ -964,9 +1011,210 @@ a result value, such as the non-result-value effects of its subexpressions.
+    of type [=ConstexprInt=] with value -2147483648.
+
+Example:  `let minint = -2147483648;` is analyzed as follows:
+* As above, `-2147483648` evaluates to a [=ConstexprInt=] value -2147483648.
+* A [=let declaration=] requires the initializer to be [=constructible=].
+    So we must apply the [=Materialization Rule=].
+* The let declaration does not have an explicit type, so we use the default materialization rule for ConstexprInt,
+    and therefore convert the ConstexprInt -2147483648 to the [=i32=] value -2147483648.
+* The result is that `minint` is declared to be the i32 value -2147483648.
+
+### Constexpr Examples ### {#constexpr-examples}
+
+<div class='example literals' heading="Type inference for literals">
+  <xmp>
+    // Explicitly-typed unsigned integer literal.
+    var u32_1 = 1u; // variable holds a u32
+
+    // Explicitly-typed signed integer literal.
+    var i32_1 = 1i; // variable holds a i32
+
+    // Explicitly-typed floating point literal.
+    var f32_1 = 1f; // variable holds a f32
+
+    // Explicitly-typed unsigned integer literal cannot be negated.
+    var u32_neg = -1u; // invalid: unary minus does not support u32
+
+    // Default inferred signed integer literal.
+    var i32_default = 1; // variable is i32
+
+    // Promotion.
+    var f32_promotion : f32 = 1; // variable holds f32
+    var i32_demotion : i32 = 1.0; // invalid, real literals cannot be demoted to integral
+
+    // Inferred from declaration type.
+    var u32_from_type : u32 = 1; // variable holds u32
+    var i32_from_type : i32 = 1; // variable holds i32
+
+    // Inferred from expression.
+    var u32_from_expr = 1 + u32_1; // variable holds u32
+    var i32_from_expr = 1 + i32_1; // variable holds i32
+
+    // Values must be representable.
+    let u32_too_large   : u32 = 1234567890123456890; // invalid, overflow
+    let i32_too_large   : i32 = 1234567890123456890; // invalid, overflow
+    let u32_large : u32 = 2147483649; // valid
+    let i32_large : i32 = 2147483649; // invalid, overflow
+    let f32_out_of_range1 = 0x1p500; // invalid, out of range
+    let f32_hex_lost_bits = 0x1.0000000001p0; // invalid, not exactly representable in f32
+
+    // Minimum integer: unary negation over ConstexprInt, then default materialization
+    // to i32.
+    let i32_min = -2147483648;  // has type i32
+    let i32_too_large_2 = 2147483648; // invalid, cannot default-materialize to i32
+
+    // TODO: If ConstexprInt gains support for addition, then these will become valid.
+    // For now they are invalid because the 1 terms default-materialize to i32,
+    // but addition does not support adding operands of different types.
+    var u32_expr1 = (1 + (1 + (1 + (1 + 1)))) + 1u; // invalid for now
+    var u32_expr2 = 1u + (1 + (1 + (1 + (1 + 1)))); // invalid for now
+    var u32_expr3 = (1 + (1 + (1 + (1u + 1)))) + 1; // invalid for now
+    var u32_expr4 = 1 + (1 + (1 + (1 + (1u + 1)))); // invalid for now
```

Comment:
What feels missing is defining what level of expression the rules are applied to. Does it only occur at a top level expression?

Reply:
Right.  I goofed. 
I think by the rules as written:    the expr3 and expr4 variants should work, with the innermost expression resolving to 1u + 1u and then cascading out to be valid.
expr1 and expr2 would be invalid because we've materialized the parenthesized expressions to i32 type.

But yes, we definitely want to get to a place where we support enough arithmetic over the special constexpr/ideal types so that all 4 will work.